### ISSUE TITLE
Add 'contents' to TaskParameter

### ISSUE THREAD

**Issue by kellrott** [ga4gh/task-execution-schemas#11]
Allow for small scripts to be injected into the system, without having to go through the object store, using a 'contents' field in the TaskParameter (and leaving the location field blank)

Example:

```
{ 
   "inputs" : [
       { "path" : "/tmp/runtime.sh", 
          "contents" : "#!/bin/bash\necho \"What up dog?\"\n"
       }
    ],
    "docker" : [
        {  
            "imageName" : "ubuntu",
        "cmd" : ["/bin/bash", "/tmp/runtime.sh"]
        }
    ]
}
```

**Comment by briandoconnor** [comment]
-1 I don't think I'd use this... I'm most concerned about large file inputs which I have to upload to an object store.  So I can't really avoid using an object store and would just bake in scripts into my docker image.

**Comment by buchanae** [comment]
I'm 50/50 on this.

I feel like client libraries will already have code to support communication with a filesystem (e.g. object storage), so this probably isn't absolutely necessary and would add one more item to the list of things needed to implement a TES server.

Also, sometimes it's nice for debugging when files like this are written to an accessible filesystem, so that you can go hunting around afterwards to understand what happened. If they are written only to the worker node, you might have less/no access to those files for debugging.

Possibly this adds security risk as well, since we're talking about writing a file that's going to get executed and we need to apply limits to the size.

But, on the other hand, it would possibly be easy to implement and might be handy for keeping temporary files out of the object store. And, you might need just as much logic to figure out how/where/when you're going to store/cleanup run scripts in the object store.

**Comment by buchanae** [comment]
This made me think of extensions:

Some APIs allow for extensions (I think Kubernetes does this) which is often a useful way to (dis)prove the utility of features before accepting into the core. Might be something keep in mind for the longer future of TES.

**Comment by geoffjentry** [comment]
I wouldn't object to this but I also feel it's necessary either. Call that less "50/50" and more "I don't care enough to argue either way" :)

**Comment by buchanae** [comment]
I've been chewing on a related idea, which is to give TES a limited storage API which would include get, put, and stat operations. TES already has a concept of storage backends, configuration, and URLs such as `file:///path/to/foo` and `s3://bucket/path/to/foo`.

`get` and `put` are obvious, I presume. `stat` would return metadata about the file.

Exposing that storage as a simple service could remove another chunk of commonly duplicated code from workflow engines and task submitters, which seems to fit the mission well. Most engines will need to write and upload a wrapper shell script, and I think they need to check for output files as well. Could a simple storage API enable both of those cases?

The API would have conservative size limits. I have thought of an exact number, but less than 32KB seems fine.

I envision that as a separate concern from a task execution schema. Perhaps that's a minor organizational detail, but I think it would be good for specs to have a separate experimental or incubation phase before they get merged into the core.

**Comment by geoffjentry** [comment]
@buchanae I mentioned this on a previous GA4GH call but I'd love to see a standardized GA4GH storage API built around the S3 schema like you propose. By that I mean how to interact, put, get files and not storing arbitrary info/metadata/etc. Would such a thing cover what you're looking for?

**Comment by buchanae** [comment]
That sounds pretty similar.

Could you clarify "and not storing arbitrary info/metadata/etc"?

**Comment by geoffjentry** [comment]
@buchanae What I mean is I basically want it to operate like a filesystem/object store and nothing more. When I brought it up on the call the convo quickly devolved (from my perspective) into people talking about storing all kinds of information along with the files.

**Comment by buchanae** [comment]
Makes sense. Yes, I agree it should be a simple as possible, and storing arbitrary metadata is more complex than I was envisioning.

One consideration is that if it's not useful *enough* and every client ends up needing a direct communication channel to do what they need, such as S3 metadata access, then the usefulness of a storage proxy API is gone.

**Comment by geoffjentry** [comment]
@buchanae Right. My thought is just that if it's not something which would fit in what the world already pictures as an object store, then that concept should be *something else* instead of trying to jam the two concepts together

**Comment by buchanae** [comment]
By the way, on Funnel at least, scripts can be easily packed into the `cmd` field
```
{
    "name": "Hello world",
    "projectId": "Funnel",
    "description": "Simple Echo Command",
    "resources": {},
    "docker": [
        {
            "imageName": "ubuntu",
            "cmd": ["/bin/bash", "-c", "#!/bin/bash\necho \"What up dog?\"\n"],
            "stdout": "stdout"
        }
    ]
}
```

If that fulfills the need, I'm not sure it's worth adding something new to the spec.

**Comment by buchanae** [comment]
To argue with myself... :)

One place where contents might be useful is in the rabix/bunny TES backend, where they need to transmit a JSON file describing the CWL job. This is about 3-10KB in my experience, and might not fit well into the `cmd` field.

**Comment by kellrott** [comment]
Fixed in #34
